### ISSUE TITLE
Failed hlstool with specific function.

### ISSUE THREAD

**Issue by HahaLan97** [llvm/circt#4719]
Hi! I'm playing around with the hlstool within this project. But I got a weird error.

I'm trying to HLS a Gaussian Filter function within CIRCT.

```
#include <ctype.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define COLS      20
#define ROWS      20
#define KERN_SIZE 3
#define KERN_RAD  (KERN_SIZE >> 1)
#define PIXELS    COLS* ROWS
#define MAX(x, y) (((x) > (y)) ? (x) : (y))
#define MIN(x, y) (((x) < (y)) ? (x) : (y))

void gauss(uint8_t input[PIXELS], uint8_t output[PIXELS])
{
    // Gauss kernel
    int16_t kernel[KERN_SIZE * KERN_SIZE] = {1, 2, 1, 2, 4, 2, 1, 2, 1};

    for (int32_t y = 0; y < ROWS; y++) {
        for (int32_t x = 0; x < COLS; x++) {
            // temp var to store the result
            int16_t sum = 0;
            for (int32_t i = -KERN_RAD; i <= KERN_RAD; i++) {
                for (int32_t j = -KERN_RAD; j <= KERN_RAD; j++) {
                    int32_t yi = y + i;
                    int32_t xj = x + j;
                    if (xj >= 0 && xj < COLS && yi >= 0 && yi < ROWS)
                        sum +=
                            (int16_t)(input[yi * COLS + xj] * kernel[(i + KERN_RAD) * KERN_SIZE + j + KERN_RAD]);
                }
            }
            // Normalize and store
            output[y * COLS + x] = (uint8_t)(MAX(MIN(255, sum >> 4), 0));
        }
    }
}
```

I firstly use `Polygeist` to translate the above C code into MLIR as following:

```
module {
  func.func @gauss(%arg0: memref<400xi8>, %arg1: memref<400xi8>) {
    %c2 = arith.constant 2 : index
    %c-1 = arith.constant -1 : index
    %c20 = arith.constant 20 : index
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %true = arith.constant true
    %c4_i32 = arith.constant 4 : i32
    %c255_i32 = arith.constant 255 : i32
    %c1_i32 = arith.constant 1 : i32
    %c3_i32 = arith.constant 3 : i32
    %c0_i16 = arith.constant 0 : i16
    %c20_i32 = arith.constant 20 : i32
    %c0_i32 = arith.constant 0 : i32
    %c4_i16 = arith.constant 4 : i16
    %c2_i16 = arith.constant 2 : i16
    %c1_i16 = arith.constant 1 : i16
    %alloca = memref.alloca() : memref<9xi16>
    affine.store %c1_i16, %alloca[0] : memref<9xi16>
    affine.store %c2_i16, %alloca[1] : memref<9xi16>
    affine.store %c1_i16, %alloca[2] : memref<9xi16>
    affine.store %c2_i16, %alloca[3] : memref<9xi16>
    affine.store %c4_i16, %alloca[4] : memref<9xi16>
    affine.store %c2_i16, %alloca[5] : memref<9xi16>
    affine.store %c1_i16, %alloca[6] : memref<9xi16>
    affine.store %c2_i16, %alloca[7] : memref<9xi16>
    affine.store %c1_i16, %alloca[8] : memref<9xi16>
    scf.for %arg2 = %c0 to %c20 step %c1 {
      %0 = arith.index_cast %arg2 : index to i32
      %1 = arith.muli %0, %c20_i32 : i32
      scf.for %arg3 = %c0 to %c20 step %c1 {
        %2 = arith.index_cast %arg3 : index to i32
        %3 = scf.for %arg4 = %c-1 to %c2 step %c1 iter_args(%arg5 = %c0_i16) -> (i16) {
          %12 = arith.index_cast %arg4 : index to i32
          %13 = arith.addi %0, %12 : i32
          %14 = arith.cmpi sge, %13, %c0_i32 : i32
          %15 = arith.cmpi slt, %13, %c20_i32 : i32
          %16 = arith.andi %14, %15 : i1
          %17 = arith.muli %13, %c20_i32 : i32
          %18 = arith.addi %12, %c1_i32 : i32
          %19 = arith.muli %18, %c3_i32 : i32
          %20 = scf.for %arg6 = %c-1 to %c2 step %c1 iter_args(%arg7 = %arg5) -> (i16) {
            %21 = arith.index_cast %arg6 : index to i32
            %22 = arith.addi %2, %21 : i32
            %23 = arith.cmpi sge, %22, %c0_i32 : i32
            %24 = arith.cmpi slt, %22, %c20_i32 : i32
            %25 = arith.andi %24, %16 : i1
            %26 = arith.andi %23, %25 : i1
            %27 = scf.if %26 -> (i16) {
              %28 = arith.addi %17, %22 : i32
              %29 = arith.index_cast %28 : i32 to index
              %30 = memref.load %arg0[%29] : memref<400xi8>
              %31 = arith.extsi %30 : i8 to i32
              %32 = arith.addi %19, %21 : i32
              %33 = arith.addi %32, %c1_i32 : i32
              %34 = arith.index_cast %33 : i32 to index
              %35 = memref.load %alloca[%34] : memref<9xi16>
              %36 = arith.extsi %35 : i16 to i32
              %37 = arith.muli %31, %36 : i32
              %38 = arith.trunci %37 : i32 to i16
              %39 = arith.addi %arg7, %38 : i16
              scf.yield %39 : i16
            } else {
              scf.yield %arg7 : i16
            }
            scf.yield %27 : i16
          }
          scf.yield %20 : i16
        }
        %4 = arith.addi %1, %2 : i32
        %5 = arith.index_cast %4 : i32 to index
        %6 = arith.extsi %3 : i16 to i32
        %7 = arith.shrsi %6, %c4_i32 : i32
        %8 = arith.cmpi sgt, %7, %c255_i32 : i32
        %result = scf.if %8 -> (i32) {
          scf.yield %c255_i32 : i32
        } else {
          scf.yield %7 : i32
        }
        %9 = scf.if %8 -> (i1) {
          scf.yield %true : i1
        } else {
          %12 = arith.cmpi sgt, %7, %c0_i32 : i32
          scf.yield %12 : i1
        }
        %10 = scf.if %9 -> (i32) {
          // %12 = arith.select %8, %c255_i32, %7 : i32
          // scf.yield %12 : i32
          scf.yield %result : i32
        } else {
          scf.yield %c0_i32 : i32
        }
        %11 = arith.trunci %10 : i32 to i8
        memref.store %11, %arg1[%5] : memref<400xi8>
      }
    }
    return
  }
}
```

Then I use the `hlstool` as:

```
hlstool gauss.mlir -dynamic-hw
```

Then the error showed up:

```
hlstool: /home/bi/Desktop/circt/llvm/mlir/lib/IR/PatternMatch.cpp:276: virtual void mlir::RewriterBase::eraseOp(mlir::Operation *): Assertion `op->use_empty() && "expected 'op' to have no uses"' failed.
PLEASE submit a bug report to https://github.com/llvm/circt and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/bi/Desktop/circt/build/bin/hlstool gauss.mlir -dynamic-hw
 #0 0x00005600d347eb1d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /home/bi/Desktop/circt/llvm/llvm/lib/Support/Unix/Signals.inc:567:11
 #1 0x00005600d347efab PrintStackTraceSignalHandler(void*) /home/bi/Desktop/circt/llvm/llvm/lib/Support/Unix/Signals.inc:641:1
 #2 0x00005600d347d326 llvm::sys::RunSignalHandlers() /home/bi/Desktop/circt/llvm/llvm/lib/Support/Signals.cpp:104:5
 #3 0x00005600d347f6d5 SignalHandler(int) /home/bi/Desktop/circt/llvm/llvm/lib/Support/Unix/Signals.inc:412:1
 #4 0x00007f5a4e6f5420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #5 0x00007f5a4e18800b raise /build/glibc-SzIz7B/glibc-2.31/signal/../sysdeps/unix/sysv/linux/raise.c:51:1
 #6 0x00007f5a4e167859 abort /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81:7
 #7 0x00007f5a4e167729 get_sysdep_segment_value /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:509:8
 #8 0x00007f5a4e167729 _nl_load_domain /build/glibc-SzIz7B/glibc-2.31/intl/loadmsgcat.c:970:34
 #9 0x00007f5a4e178fd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
#10 0x00005600d4115a34 mlir::RewriterBase::eraseOp(mlir::Operation*) /home/bi/Desktop/circt/llvm/mlir/lib/IR/PatternMatch.cpp:0:3
#11 0x00005600d3ad385a (anonymous namespace)::EliminateCBranchIntoMuxPattern::matchAndRewrite(circt::handshake::MuxOp, mlir::PatternRewriter&) const /home/bi/Desktop/circt/lib/Dialect/Handshake/HandshakeOps.cpp:349:12
#12 0x00005600d3ae74cb mlir::detail::OpOrInterfaceRewritePatternBase<circt::handshake::MuxOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const /home/bi/Desktop/circt/llvm/mlir/include/mlir/IR/PatternMatch.h:330:12
#13 0x00005600d47b1810 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) /home/bi/Desktop/circt/llvm/mlir/lib/Rewrite/PatternApplicator.cpp:200:25
#14 0x00005600d473dd53 (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) /home/bi/Desktop/circt/llvm/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:293:19
#15 0x00005600d473d1ee mlir::applyPatternsAndFoldGreedily(llvm::MutableArrayRef<mlir::Region>, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) /home/bi/Desktop/circt/llvm/mlir/lib/Transforms/Utils/GreedyPatternRewriteDriver.cpp:453:8
#16 0x00005600d428b005 mlir::applyPatternsAndFoldGreedily(mlir::Operation*, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) /home/bi/Desktop/circt/llvm/mlir/include/mlir/Transforms/GreedyPatternRewriteDriver.h:86:10
#17 0x00005600d4288cbc (anonymous namespace)::Canonicalizer::runOnOperation() /home/bi/Desktop/circt/llvm/mlir/lib/Transforms/Canonicalizer.cpp:61:9
#18 0x00005600d481a99a mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:471:21
#19 0x00005600d481afb4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:534:16
#20 0x00005600d4820598 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::$_14::operator()(mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo&) const /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:754:36
#21 0x00005600d4820209 mlir::LogicalResult mlir::failableParallelForEach<__gnu_cxx::__normal_iterator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo*, std::vector<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo, std::allocator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo>>>, mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::$_14&>(mlir::MLIRContext*, __gnu_cxx::__normal_iterator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo*, std::vector<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo, std::allocator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo>>>, __gnu_cxx::__normal_iterator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo*, std::vector<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo, std::allocator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo>>>, mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::$_14&) /home/bi/Desktop/circt/llvm/mlir/include/mlir/IR/Threading.h:46:18
#22 0x00005600d481c293 mlir::LogicalResult mlir::failableParallelForEach<std::vector<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo, std::allocator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo>>&, mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::$_14&>(mlir::MLIRContext*, std::vector<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo, std::allocator<mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::OpPMInfo>>&, mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool)::$_14&) /home/bi/Desktop/circt/llvm/mlir/include/mlir/IR/Threading.h:92:10
#23 0x00005600d481bb37 mlir::detail::OpToOpPassAdaptor::runOnOperationAsyncImpl(bool) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:764:14
#24 0x00005600d481ac67 mlir::detail::OpToOpPassAdaptor::runOnOperation(bool) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:655:5
#25 0x00005600d481a98b mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:468:5
#26 0x00005600d481afb4 mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:534:16
#27 0x00005600d481c9dc mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:843:10
#28 0x00005600d481c8f2 mlir::PassManager::run(mlir::Operation*) /home/bi/Desktop/circt/llvm/mlir/lib/Pass/Pass.cpp:823:60
#29 0x00005600d3354107 doHLSFlowDynamic(mlir::PassManager&, mlir::ModuleOp, std::optional<std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>>&) /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:385:17
#30 0x00005600d3353bce processBuffer(mlir::MLIRContext&, mlir::TimingScope&, llvm::SourceMgr&, std::optional<std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>>&) /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:427:16
#31 0x00005600d3353826 processInputSplit(mlir::MLIRContext&, mlir::TimingScope&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, std::optional<std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>>&) /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:449:12
#32 0x00005600d335369d processInput(mlir::MLIRContext&, mlir::TimingScope&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, std::optional<std::unique_ptr<llvm::ToolOutputFile, std::default_delete<llvm::ToolOutputFile>>>&) /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:465:12
#33 0x00005600d3353554 executeHlstool(mlir::MLIRContext&) /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:500:14
#34 0x00005600d3353356 main /home/bi/Desktop/circt/tools/hlstool/hlstool.cpp:556:17
#35 0x00007f5a4e169083 __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:342:3
#36 0x00005600d335313e _start (/home/bi/Desktop/circt/build/bin/hlstool+0xe1d13e)
[1]    5147 abort (core dumped)  /home/bi/Desktop/circt/build/bin/hlstool gauss.mlir -dynamic-hw
```

Which I've asked about once on discourse, this should be a bug somewhere in the lowering.

**Comment by youngar** [comment]
Crash is probably in this part of the code: https://github.com/llvm/circt/blob/55e2230e3fd12d9ce43d825fe5242a2a3899e9c3/lib/Dialect/Handshake/HandshakeOps.cpp#L341-L344

It is super weird to have the call to `rewriter.replaceOp` wrapped in `rewriter.updateRootInPlace`, and even weirder that they are changing a different operation then the one we are supposedly transacting on.  The whole thing can probably be replaced with just:
```c++
rewriter.replaceOp(op, firstParentCBranch.getDataOperand()); 
```

**Comment by youngar** [comment]
Looks like the issue was actually fixed recently: https://github.com/llvm/circt/commit/083de301f0affa73d2c7f59ea73dbd03acd57bb6

**Comment by HahaLan97** [comment]
> Looks like the issue was actually fixed recently: [083de30](https://github.com/llvm/circt/commit/083de301f0affa73d2c7f59ea73dbd03acd57bb6)

Oh! I'll try this later!

**Comment by HahaLan97** [comment]
> Looks like the issue was actually fixed recently: [083de30](https://github.com/llvm/circt/commit/083de301f0affa73d2c7f59ea73dbd03acd57bb6)

I've pulled and rebuilt the project. This problem is already fixed! Thanks for pointing out, I'll close this issue:)
